include/abb_librws/common/rw/elog.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <chrono>
+#include <variant>
+
+namespace abb ::rws ::rw ::elog
+{
+
+    /**
+     * @brief Types of elog messages
+     */
+    enum class ElogMessageType
+    {
+        INFORMATION = 1,
+        WARNING = 2,
+        ERROR = 3
+    };
+
+    std::string to_string(ElogMessageType const messagType);
+
+
+    /**
+     * @brief Argument of elog message.
+     * Can be of string, long or float type.
+     * https://developercenter.robotstudio.com/api/rwsApi/elog_domain_get_page.html
+     */
+    using ElogMessageArg = std::variant<std::string, long, float>;
+
+    ElogMessageArg makeElogMessageArg(std::string const &type, std::string const& value);
+
+    std::string to_string(ElogMessageArg const& arg);
+
+
+    /**
+     * @brief Elog message structure
+     * https://developercenter.robotstudio.com/api/rwsApi/elog_domain_get_page.html
+     */
+    struct ElogMessage
+    {
+        int domain;                       // Domain of message.
+        int sequenceNumber;               // Sequence number of message in domain.
+        ElogMessageType messageType;          // The elog message type.
+        int code;                         // The elog message code.
+        std::string sourceName;           // The elog message source.
+        std::tm timestamp;                // The time stamp when the event log was generated.
+        std::string title;                // The elog message title in the specified langauge. Available only when lang parameter is provided.
+        std::string desc;                 // The elog message description in the specified langauge. Available only when lang parameter is provided.
+        std::string conseqs;              // The elog message consequences in the specified langauge. Available only when lang parameter is provided.
+        std::string causes;               // The elog message causes in the specified langauge. Available only when lang parameter is provided.
+        std::string actions;              // The elog message actions in the specified langauge. Available only when lang parameter is provided.
+        
+        /**
+         * @brief Message arguments necessary for filling message templates.
+         * If valid language is provided when retrieving /rw/elog message, args are filled in appropirate places. 
+         * @param first arg name: e.g. arg1, arg2 etc.. 
```

Comment:
Are the arguments identified by their order (0, 1, 2, ...) or by names? If by order, the previous version with `std::vector` might be more convenient.

Reply:
The message templates are stored in XML files in /fileservice/products/RobotControl_7.2.2/language/en/controller/elog/` directory in format:
```
  <Message number="1816" eDefine="SYS_ERR_ARL_TOO_MANY_CONNECTED_CYCLIC">
    <Title>Too many connected Cyclic bool</Title>
    <Description>Task: <arg format="%s" ordinal="1" />.<p />It is only allowed to setup <arg format="%s" ordinal="4" /> number of cyclic bools.</Description>
    <Consequences>The program execution is stopped immediately. The given condition will not be connected to <arg format="%s" ordinal="2" />.</Consequences>
    <Causes>The maximum number of cyclic bools (<arg format="%s" ordinal="4" />) has already been setup.</Causes>
    <Actions>Remove all cyclic bools not used for the moment and try again.</Actions>
  </Message>
```
The `ordinal` (indexing from 1) attribute of `arg` tag indicates the argument that should be used in that place.
Maybe `std::map<int, ElogMessageArg>` would be more appropriate.